File: 199-binary-tree-right-side-view/199-binary-tree-right-side-view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int max_level = -1;
    private void rightView(TreeNode root,int level,List<Integer> result){
        if(root == null) return;
        
        if(level > max_level){
            result.add(root.val);
            max_level = level;
        }
        rightView(root.right,level + 1,result);
        rightView(root.left,level + 1,result);
    }
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        rightView(root,0,result);
        return result;
    }
}

File: 199-binary-tree-right-side-view/199-binary-tree-right-side-view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void solve(TreeNode *root,int depth,int &maxDepth,vector<int> &rightview) {
        if(root == NULL) {
            return;
        }
        
        if(depth > maxDepth) {
            maxDepth = depth;
            rightview.push_back(root->val);
        }
        
        solve(root->right,depth + 1,maxDepth,rightview);
        solve(root->left,depth + 1,maxDepth,rightview);
    }
    vector<int> rightSideView(TreeNode* root) {
        int depth = 0;
        int maxDepth = -1;
        
        vector<int> rightview;
        solve(root,depth,maxDepth,rightview);
        return rightview;
    }
};